buzzwindow.cpp

```diff
@@ -73,11 +74,19 @@ void BuzzWindow::MessageReceived(BMessage* message)
 				Entry.GetPath(&path);
 				if((Entry.GetRef(&ref) == B_OK))
 				{
-					MyClock->SetSound(new BFileGameSound(&ref, false));
-					SoundName->SetText(ref.name);
+					if (BFileGameSound(&ref, false).InitCheck() == B_OK) {
+						MyClock->SetSound(new BFileGameSound(&ref, false));
+						SoundName->SetText(ref.name);
+					} else {
+						BAlert* not_supported = new BAlert(
+							"Error",
+							"This is not a supported sound file.",
+							"OK", NULL, NULL, B_WIDTH_AS_USUAL,
+							B_OFFSET_SPACING, B_WARNING_ALERT);
+						not_supported->Go();
+					}
 				}
 			}
-   		
    			else BWindow::MessageReceived(message);break;
```

Comment:
How does the logic need to change? It seems to work well...

Reply:
BFileGameSound is instanciated twice.

Reply:
Fixed.

Reply:
> How does the logic need to change? It seems to work well...

Writing efficient, self-explanatory (whenever possible), and robust code is much more than "It seems to work well..." :)

In addition to the problem that `BFileGameSound` was instantiated twice (as @korli pointed out), the if-statement `if((Entry.GetRef(&ref) == B_OK))` is superfluous because `Entry` was just initialized from `ref` without symlink traversing. Also, the local variable `path` was not used in the `case` block and thus should be removed. Finally, the nested if-else statements are unnecessarily deep.

So you can do something like the following:
```
			entry_ref ref;
			if (message->FindRef("refs", &ref) != B_OK) {
				BWindow::MessageReceived(message);
				break;
			}

			BFileGameSound soundFile(&ref, false);
				// The above line is wrong. Use pointer/new instead.
			if (soundFile.InitCheck() != B_OK) {
				BAlert* alert = new BAlert("Error",
					"This is not a supported sound file.", "OK");
				alert->SetType(B_WARNING_ALERT);
				alert->Go();
				break;
			}

			MyClock->SetSound(&soundFile);
			SoundName->SetText(ref.name);
```

Reply:
Thank you for your explanation. I am still finding my way around C++, and I am training my eye to judge between bad and good code. I've reworked the logic as suggested.